SegmentStore: Segment deletion race between ReadIndex and StorageWriter

**Problem description**
When a SegmentContainer deletes a segment it does the following steps:
1. Delete from Metadata
2. Delete state files
3. Remove Read Indices

The StorageWriter will eventually pick up deleted Segments and remove their associated `SegmentAggregators` (and it has plenty of checks to see if the Segment was deleted in the meantime). However, there is a possibility that `#3` above happens after the last check for `IsDeleted` but before the actual attempt to read data for a potential flush operation. In this case, the `SegmentAggregator` would incorrectly throw a `DataCorruptionException` since it cannot find the associated data in the ReadIndex.

Example error stack trace:
```
 java.util.concurrent.CompletionException: io.pravega.segmentstore.server.DataCorruptionException: Unable to retrieve CacheContents for 'AggregatedAppend: SegmentId = 28, Offsets = [0-91170), SeqNo = 48'.
    ...
    Caused by: io.pravega.segmentstore.server.DataCorruptionException: Unable to retrieve CacheContents for 'AggregatedAppend: SegmentId = 28, Offsets = [0-91170), SeqNo = 48'.
    	at io.pravega.segmentstore.server.writer.SegmentAggregator.getFlushArgs(SegmentAggregator.java:651)
    	at io.pravega.segmentstore.server.writer.SegmentAggregator.flushPendingAppends(SegmentAggregator.java:598)
    	at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$flushFully$11(SegmentAggregator.java:557)
    	at io.pravega.common.concurrent.FutureHelpers$Loop.call(FutureHelpers.java:623)
    	... 10 common frames omitted
```
**Problem location**
SegmentAggregator.java

**Suggestions for an improvement**
Add a check in `SegmentAggregator.java:getFlushArgs` to not throw if the read result is null and the Segment is deleted.

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java
@@ -575,7 +575,7 @@ class SegmentAggregator implements OperationProcessor, AutoCloseable {
         FlushResult result = new FlushResult();
         return FutureHelpers
                 .loop(
-                        this::exceedsThresholds,
+                        () -> !this.metadata.isDeleted() && exceedsThresholds(),
                         () -> flushPendingAppends(timer.getRemaining()),
                         result::withFlushResult,
                         executor)
@@ -648,6 +648,10 @@ class SegmentAggregator implements OperationProcessor, AutoCloseable {
         int length = (int) appendOp.getLength();
         InputStream data = this.dataSource.getAppendData(appendOp.getStreamSegmentId(), appendOp.getStreamSegmentOffset(), length);
         if (data == null) {
+            if (this.metadata.isDeleted()) {
+                // Segment was deleted - nothing more to do.
+                return new FlushArgs(null, 0);
+            }
             throw new DataCorruptionException(String.format("Unable to retrieve CacheContents for '%s'.", appendOp));
         }
 

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/writer/SegmentAggregatorTests.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/writer/SegmentAggregatorTests.java
@@ -973,6 +973,72 @@ public class SegmentAggregatorTests extends ThreadPooledTestSuite {
         Assert.assertArrayEquals("Unexpected data written to storage.", expectedData, actualData);
     }
 
+    /**
+     * Tests the case when a Segment is deleted in the Metadata and ReadIndex.
+     * Note that we are not testing the case when the Segment is deleted only in Storage, since the SegmentContainer
+     * first deletes in the Metadata and then in Storage, and that would simply throw a StreamSegmentNotExistsException,
+     * which will subside as soon as the Metadata is updated and the StorageWriter cleans up the SegmentAggregator.
+     */
+    @Test
+    public void testDeletedSegmentInMetadata() throws Exception {
+        final WriterConfig config = DEFAULT_CONFIG;
+
+        @Cleanup
+        TestContext context = new TestContext(config);
+        context.storage.create(context.segmentAggregator.getMetadata().getName(), TIMEOUT).join();
+        context.segmentAggregator.initialize(TIMEOUT, executorService()).join();
+
+        // Add one operation big enough to trigger a Flush.
+        byte[] appendData = new byte[config.getFlushThresholdBytes() + 1];
+        StorageOperation appendOp = generateAppendAndUpdateMetadata(SEGMENT_ID, appendData, context);
+
+        context.segmentAggregator.add(appendOp);
+
+        Assert.assertTrue("Unexpected value returned by mustFlush() (size threshold).", context.segmentAggregator.mustFlush());
+
+        //Delete the segment in the Metadata & Read index. We want to make sure we do this while the flush() method is
+        //running, hence the callback (flush() has a check at the beginning that exits if the metadata indicates deleted).
+        context.dataSource.setOnGetAppendData(() -> {
+            context.containerMetadata.deleteStreamSegment(context.segmentAggregator.getMetadata().getName());
+            context.dataSource.clearAppendData();
+        });
+
+        // Call flush() and inspect the result.
+        FlushResult flushResult = context.segmentAggregator.flush(TIMEOUT, executorService()).join();
+        Assert.assertEquals("Not expecting any bytes to be flushed.", 0, flushResult.getFlushedBytes());
+        Assert.assertEquals("Not expecting any merged bytes in this test.", 0, flushResult.getMergedBytes());
+        Assert.assertFalse("Unexpected value returned by mustFlush() after flush.", context.segmentAggregator.mustFlush());
+    }
+
+    /**
+     * Tests the case when a Segment's data is missing from the ReadIndex (but the Segment itself is not deleted).
+     */
+    @Test
+    public void testSegmentMissingData() throws Exception {
+        final WriterConfig config = DEFAULT_CONFIG;
+
+        @Cleanup
+        TestContext context = new TestContext(config);
+        context.storage.create(context.segmentAggregator.getMetadata().getName(), TIMEOUT).join();
+        context.segmentAggregator.initialize(TIMEOUT, executorService()).join();
+
+        // Add one operation big enough to trigger a Flush.
+        byte[] appendData = new byte[config.getFlushThresholdBytes() + 1];
+        StorageOperation appendOp = generateAppendAndUpdateMetadata(SEGMENT_ID, appendData, context);
+
+        context.segmentAggregator.add(appendOp);
+        Assert.assertTrue("Unexpected value returned by mustFlush() (size threshold).", context.segmentAggregator.mustFlush());
+
+        // Clear the append data.
+        context.dataSource.clearAppendData();
+
+        // Call flush() and verify it throws DataCorruptionException.
+        AssertExtensions.assertThrows(
+                "flush() did not throw when unable to read data from ReadIndex.",
+                () -> context.segmentAggregator.flush(TIMEOUT, executorService()),
+                ex -> ex instanceof DataCorruptionException);
+    }
+
     //endregion
 
     //region Unknown outcome operation reconciliation

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/writer/TestWriterDataSource.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/writer/TestWriterDataSource.java
@@ -79,6 +79,8 @@ class TestWriterDataSource implements WriterDataSource, AutoCloseable {
     private ErrorInjector<Exception> getAppendDataErrorInjector;
     @GuardedBy("lock")
     private BiConsumer<Long, Long> completeMergeCallback;
+    @GuardedBy("lock")
+    private Runnable onGetAppendData;
     private final Object lock = new Object();
 
     //endregion
@@ -259,11 +261,17 @@ class TestWriterDataSource implements WriterDataSource, AutoCloseable {
 
     @Override
     public InputStream getAppendData(long streamSegmentId, long startOffset, int length) {
+        Runnable callback;
+        AppendData ad;
         synchronized (this.lock) {
             ErrorInjector.throwSyncExceptionIfNeeded(this.getAppendDataErrorInjector);
+            callback = this.onGetAppendData;
         }
 
-        AppendData ad;
+        if (callback != null) {
+            callback.run();
+        }
+
         synchronized (this.lock) {
             ad = this.appendData.getOrDefault(streamSegmentId, null);
         }
@@ -307,6 +315,12 @@ class TestWriterDataSource implements WriterDataSource, AutoCloseable {
     //endregion
 
     //region Other Properties
+
+    void setOnGetAppendData(Runnable callback) {
+        synchronized (this.lock) {
+            this.onGetAppendData = callback;
+        }
+    }
 
     void setSegmentMetadataRequested(Consumer<Long> callback) {
         synchronized (this.lock) {

